### ISSUE TITLE
broken since 5.2.0

### ISSUE THREAD

**Issue by cfsghost** [alexmingoia/koa-router#181]
I was using v5.1.2, it works well on my project before.

but it's broken since 5.2+ which includes v5.2.1. No route matches is it.

**Comment by nickpoorman** [comment]
+1

**Comment by nickpoorman** [comment]
I can no longer set `this.someobject` and access if from the next middleware.

**Comment by alexmingoia** [comment]
Please show me a test that demonstrates this behavior. The test for properties on `this` context is passing: https://github.com/alexmingoia/koa-router/blob/master/test/lib/router.js#L294-L320

**Comment by luisfontes** [comment]
It broke for me as well, I'm getting the error below. Rolling back to 5.1.2 and everything works as usual.


```javascript
/var/app/current/node_modules/koa-router/lib/router.js:513
      var routeNestingLevel = route.path.split('/').length;
                                         ^
TypeError: undefined is not a function
    at Router.register (/var/app/current/node_modules/koa-router/lib/router.js:513:42)
    at Router.(anonymous function) [as options] (/var/app/current/node_modules/koa-router/lib/router.js:182:10)
    at Object.<anonymous> (/var/app/current/routes.js:39:4)
    at Module._compile (module.js:460:26)
    at normalLoader (/var/app/current/node_modules/babel/node_modules/babel-core/lib/api/register/node.js:199:5)
    at Object.require.extensions.(anonymous function) [as .js] (/var/app/current/node_modules/babel/node_modules/babel-core/lib/api/register/node.js:216:7)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/var/app/current/server.js:6:37)
    at Module._compile (module.js:460:26)
    at normalLoader (/var/app/current/node_modules/babel/node_modules/babel-core/lib/api/register/node.js:199:5)
    at Object.require.extensions.(anonymous function) [as .js] (/var/app/current/node_modules/babel/node_modules/babel-core/lib/api/register/node.js:216:7)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/var/app/current/index.js:4:11)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
```

**Comment by chrisveness** [comment]
I'm also getting what looks like a related issue:

      TypeError: m.path.split is not a function
          at /.../node_modules/koa-router/lib/router.js:516:36
          at Array.some (native)
          at Router.register (/.../node_modules/koa-router/lib/router.js:515:21)
          at Router.(anonymous function) [as get] (/.../node_modules/koa-router/lib/router.js:182:10)
          at Object.<anonymous> (/.../apps/www/routes-www.js:31:8)
          at Module._compile (module.js:434:26)
          at Object.Module._extensions..js (module.js:452:10)
          at Module.load (module.js:355:32)
          at Function.Module._load (module.js:310:12)
          at Module.require (module.js:365:17)

I'm afraid I don't have time to isolate a test case for you right now, but I identified the error was introduced between 5.2.1 and 5.2.2, if that helps. I am reverting to 5.2.1 for now.

**Comment by ricardo-nog** [comment]
#188 fixes this.

**Comment by ruiquelhas** [comment]
#152 adds tests and fix for the specific Regexp issue.

**Comment by chrisveness** [comment]
I believe #188 doesn't completely fix this issue: see my comment on DimensionSoftware/koa-router@b740a3d.

**Comment by ricardo-nog** [comment]
+1 on #152.

**Comment by mzgol** [comment]
Ping. Using regular expressions with koa-router is broken without this fix; #152 looks good at first sight.

**Comment by tarrencev** [comment]
Eager for this fix too.

**Comment by alexmingoia** [comment]
This is fixed in 6.x. Koa is moving to a different API with 2.x. koa-router 6 has been released which requires koa 2.

**Comment by chrisveness** [comment]
Koa v2 is still in alpha, and if I understand well, when it is released will require either a clumsier syntax or Babel transpiling for async/await, neither of which I want.

If v6+ requires Koa v2, could this fix be backported into v5?

**Comment by alexmingoia** [comment]
@chrisveness #152 was merged into `5.3.0` so it should be fixed for you on 5.x as well.

**Comment by chrisveness** [comment]
All is sweet with 5.3.0, thanks.
